src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinHttpWebSocket.cs

```diff
@@ -441,6 +448,12 @@ private Task<bool> InternalReceiveAsync(ArraySegment<byte> buffer)
                 Abort();
             }
 
+            // If the task has completed, there's no need to force continuations to run async
+            if (_operation.TcsReceive.Task.Status == TaskStatus.RanToCompletion)
+            {
+                return Task.FromResult(_operation.TcsReceive.Task.GetAwaiter().GetResult());
+            }
```

Comment:
The websocket is being wrapped into a stream and passed to a WCF message encoder. That message encoder then uses an XmlReader (XmlUTF8TextReader is the concrete type) to read xml nodes from the stream. As there are only synchronous methods for reading the Xml nodes, when more data is needed, it can only call synchronous methods on the stream. This results in effectively blocking the thread that called WebSocket.ReceiveAsync calling Wait on the resulting Task. Even if we were to fix the XmlReader to be async, we could have a problem elsewhere as none of our serializers have an async code path so passing a stream to any of them could cause this same issue. Fundamentally the issue is the WebSocket class runs a Task continuation it doesn't own on a thread it has no business running continuations on.